# Выбор локации для бурения скважины

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Анализ-и-подготовка-данных" data-toc-modified-id="Анализ-и-подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Анализ и подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Список-используемых-библиотек-и-функций" data-toc-modified-id="Список-используемых-библиотек-и-функций-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Список используемых библиотек и функций</a></span></li><li><span><a href="#Загрузка-данных-и-первичный-анализ" data-toc-modified-id="Загрузка-данных-и-первичный-анализ-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Загрузка данных и первичный анализ</a></span></li><li><span><a href="#Предобработка-данных-на-основе-анализа" data-toc-modified-id="Предобработка-данных-на-основе-анализа-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Предобработка данных на основе анализа</a></span></li><li><span><a href="#Вывод-по-блоку-&quot;Анализ-и-подготовка-данных&quot;" data-toc-modified-id="Вывод-по-блоку-&quot;Анализ-и-подготовка-данных&quot;-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span><strong>Вывод по блоку "Анализ и подготовка данных"</strong></a></span></li></ul></li><li><span><a href="#Обучение-и-проверка-модели" data-toc-modified-id="Обучение-и-проверка-модели-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение и проверка модели</a></span><ul class="toc-item"><li><span><a href="#Разделение-популяции-на-обучающую-и-валидационную-выборку" data-toc-modified-id="Разделение-популяции-на-обучающую-и-валидационную-выборку-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Разделение популяции на обучающую и валидационную выборку</a></span></li><li><span><a href="#Масштабирование-параметров" data-toc-modified-id="Масштабирование-параметров-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Масштабирование параметров</a></span></li><li><span><a href="#Обучим-модель-Линейной-регрессии" data-toc-modified-id="Обучим-модель-Линейной-регрессии-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обучим модель Линейной регрессии</a></span></li><li><span><a href="#Вывод-по-блоку-&quot;Обучение-и-проверка-модели&quot;" data-toc-modified-id="Вывод-по-блоку-&quot;Обучение-и-проверка-модели&quot;-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span><strong>Вывод по блоку "Обучение и проверка модели"</strong></a></span></li></ul></li><li><span><a href="#Подготовка-к-расчёту-прибыли" data-toc-modified-id="Подготовка-к-расчёту-прибыли-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Подготовка к расчёту прибыли</a></span><ul class="toc-item"><li><span><a href="#Константы-для-расчетов" data-toc-modified-id="Константы-для-расчетов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Константы для расчетов</a></span></li><li><span><a href="#Точка-безубыточности" data-toc-modified-id="Точка-безубыточности-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Точка безубыточности</a></span></li><li><span><a href="#Сравнение-точки-безубыточности-с-запасами-в-регионе" data-toc-modified-id="Сравнение-точки-безубыточности-с-запасами-в-регионе-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Сравнение точки безубыточности с запасами в регионе</a></span></li><li><span><a href="#Вывод-по-блоку-&quot;Подготовка-к-расчету-прибыли&quot;" data-toc-modified-id="Вывод-по-блоку-&quot;Подготовка-к-расчету-прибыли&quot;-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span><strong>Вывод по блоку "Подготовка к расчету прибыли"</strong></a></span></li></ul></li><li><span><a href="#Расчёт-прибыли-и-рисков" data-toc-modified-id="Расчёт-прибыли-и-рисков-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Расчёт прибыли и рисков</a></span><ul class="toc-item"><li><span><a href="#Функция-расчета-прибыли" data-toc-modified-id="Функция-расчета-прибыли-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Функция расчета прибыли</a></span></li><li><span><a href="#Применение-процедуры-Bootstrap" data-toc-modified-id="Применение-процедуры-Bootstrap-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Применение процедуры Bootstrap</a></span></li><li><span><a href="#Выбор-лучшего-региона" data-toc-modified-id="Выбор-лучшего-региона-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Выбор лучшего региона</a></span></li><li><span><a href="#Вывод-по-блоку-&quot;Расчет-прибыли-и-рисков&quot;" data-toc-modified-id="Вывод-по-блоку-&quot;Расчет-прибыли-и-рисков&quot;-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span><strong>Вывод по блоку "Расчет прибыли и рисков"</strong></a></span></li></ul></li><li><span><a href="#Вывод-по-проекту" data-toc-modified-id="Вывод-по-проекту-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод по проекту</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

**Описание прокта:**

Для нефтедобывающей компании "ГлавРосГосНефть" необходимо принять решение о местоположении бурения скважины.

Компания провела исследование месторождений в трёх регионах, для каждого региона было выбрано 10 000 месторождений нефти. Исследование включало получение информации в отношении качества нефти и объемов запасов месторождения. На основе исследования были предоставлены данные о пробах в трёх исследуемых регионах.


**Цель работы:**

Необходимо построить модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализировать возможную прибыль и риски техникой Bootstrap.

**Вводные данные для определения прибыльности:**

Имеются следующие данные на основе которых можно строить понимание о прибыльности/убыточности месторождения:
- Для обучения модели подходит только линейная регрессия (остальные — недостаточно предсказуемые).
- При разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки.
- Бюджет на разработку скважин в регионе — 10 млрд рублей.
- При нынешних ценах один баррель сырья приносит 450 рублей дохода. Доход с каждой единицы продукта составляет 450 тыс. рублей, поскольку объём указан в тысячах баррелей.
- После оценки рисков нужно оставить лишь те регионы, в которых вероятность убытков меньше 2.5%. Среди них выбирают регион с наибольшей средней прибылью.

## Анализ и подготовка данных

В рамках данного блока будут проведены следующие операции:
- формирование списка используемых библиотек;
- загрузка данных и первичный анализ;
- предобработка данных на основе анализа

### Список используемых библиотек и функций

Ниже представлены все импортируемые библиотеки, которые используются в рамках данного проекта.

In [1]:
import pandas as pd
import scipy as st
import numpy as np 
import matplotlib.pyplot as plt
from sklearn.metrics import mean_squared_error

from tqdm import tqdm
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler 

### Загрузка данных и первичный анализ

Нам было предоставлено три датасета в которых имеется информация о 10 000 месторождениях для каждого региона. Загрузим информацию из файлов и проведем первичный анализ

In [2]:
#загрузим датасеты

data_1 = pd.read_csv ('/datasets/geo_data_0.csv')
data_2 = pd.read_csv ('/datasets/geo_data_1.csv')
data_3 = pd.read_csv ('/datasets/geo_data_2.csv')

In [3]:
#Проведем анализ первого датасета

print('Общая информация о данных:','\n')
print(data_1.info(),'\n')

print ('Количество дубликатов:',data_1[data_1['id'].duplicated()]['id'].count())

data_1.head(15)

Общая информация о данных: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None 

Количество дубликатов: 10


id        f0        f1        f2     product
0   txEyH  0.705745 -0.497823  1.221170  105.280062
1   2acmU  1.334711 -0.340164  4.365080   73.037750
2   409Wp  1.022732  0.151990  1.419926   85.265647
3   iJLyR -0.032172  0.139033  2.978566  168.620776
4   Xdl7t  1.988431  0.155413  4.751769  154.036647
5   wX4Hy  0.969570  0.489775 -0.735383   64.741541
6   tL6pL  0.645075  0.530656  1.780266   49.055285
7   BYPU6 -0.400648  0.808337 -5.624670   72.943292
8   j9Oui  0.643105 -0.551583  2.372141  113.356160
9   OLuZU  2.173381  0.563698  9.441852  127.910945
10  b8WQ6  0.371066 -0.036585  0.009208   70.326617
11  1YYm1  0.015920  1.062729 -0.722248   45.110381
12  zIYPq -0.276476  0.924865  0.095584   89.158678
13  iqTqq  0.212696 -0.111147  5.770095  164.298520
14  Ct5yY -0.018578  0.187516  2.944683  158.633720

In [4]:
#Проведем анализ первого датасета

print('Общая информация о данных:','\n')
print(data_2.info(),'\n')

print ('Количество дубликатов:', data_2[data_2['id'].duplicated()]['id'].count())

data_2.head(15)

Общая информация о данных: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None 

Количество дубликатов: 4


id         f0         f1        f2     product
0   kBEdx -15.001348  -8.276000 -0.005876    3.179103
1   62mP7  14.272088  -3.475083  0.999183   26.953261
2   vyE1P   6.263187  -5.948386  5.001160  134.766305
3   KcrkZ -13.081196 -11.506057  4.999415  137.945408
4   AHL4O  12.702195  -8.147433  5.004363  134.766305
5   HHckp  -3.327590  -2.205276  3.003647   84.038886
6   h5Ujo -11.142655 -10.133399  4.002382  110.992147
7   muH9x   4.234715  -0.001354  2.004588   53.906522
8   YiRkx  13.355129  -0.332068  4.998647  134.766305
9   jG6Gi   1.069227 -11.025667  4.997844  137.945408
10  7uHUe  11.777049  -5.334084  2.003033   53.906522
11  OXyvW  16.320755  -0.562946 -0.001783    0.000000
12  cB8re   7.736313  -6.093374  3.982531  107.813044
13  igmai   6.695604  -0.749449 -0.007630    0.000000
14  3OdkE -10.985487  -5.605994  2.991130   84.038886

In [5]:
#Проведем анализ первого датасета

print('Общая информация о данных:','\n')
print(data_3.info(),'\n')

print ('Количество дубликатов:',data_3[data_3['id'].duplicated()]['id'].count())

data_3.head(15)

Общая информация о данных: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None 

Количество дубликатов: 4


id        f0        f1        f2     product
0   fwXo0 -1.146987  0.963328 -0.828965   27.758673
1   WJtFt  0.262778  0.269839 -2.530187   56.069697
2   ovLUW  0.194587  0.289035 -5.586433   62.871910
3   q6cA6  2.236060 -0.553760  0.930038  114.572842
4   WPMUX -0.515993  1.716266  5.899011  149.600746
5   LzZXx -0.758092  0.710691  2.585887   90.222465
6   WBHRv -0.574891  0.317727  1.773745   45.641478
7   XO8fn -1.906649 -2.458350 -0.177097   72.480640
8   ybmQ5  1.776292 -0.279356  3.004156  106.616832
9   OilcN -1.214452 -0.439314  5.922514   52.954532
10  myq2g  0.628659 -2.120049  5.362437  166.799132
11  V9R1y -0.544565  0.039088  8.635293  150.859435
12  bKDo0 -0.960453 -0.127113  7.764734   96.910855
13  eXvhs  1.432891 -1.242074  3.557168    8.704162
14  wI4j3 -0.935464  2.420523  1.150536  116.829811

**Выводы по пункту 1.2:**

При анализе полученных данных можно выделить следующие ключевые моменты:
- в каждом из следующих датасетов представлено 10 000 строк и 5 столбцов;
- пустые значения отсутствуют;
- данные в каждом датасете соответствуют типу друг друга;
- наименование колонок в строчной записи;
- в данных имеются дубликаты по показателю `id`

Датафреймы имеют следующие признаки:

- `id` — уникальный идентификатор скважины;
- `f0`, `f1`, `f2` — три признака точек (синтетические параметры, важные для обучения);
- `product` — объём запасов в скважине (тыс. баррелей).

Таким образом можно выделить следующие предпосылки по предобработке данных:
1. При обучении модели признак `id` не нужен, он никак не влияет на прогнозирование модели, таким образом для обучения его необходимо удалить.
2. Так как уникальным идентификатором месторождения является `id`, и при этом в популяции `id` месторождения могут повторяться - необходимо сформировать список дубликатов и на финальном этапе убедиться что выбранное для бурения месторождение не попадает в список дубликатов. Таким образом мы избежим ошибку в случае если наиболее прибыльным месторождением окажется месторождение с дублирующимися `id`. 

### Предобработка данных на основе анализа

На основе проведенного анализа данных выделены следующие шаги по предобработке данных:

1. выделение списка дубликатов в признаке `id` в отдельный массив;
2. удаления признака `id` из датасета для обучения модели.

In [6]:
#сформируем список дубликатов для каждого региона

duplicated_list_1 = data_1[data_1['id'].duplicated()]['id']
duplicated_list_2 = data_2[data_2['id'].duplicated()]['id']
duplicated_list_3 = data_3[data_2['id'].duplicated()]['id']

#выведем полученную информацию на экран

print('Количество дубликатов для первого региона:',len(duplicated_list_1),'\n'
      'Список дубликатов для первого региона:','\n',duplicated_list_1,'\n','\n'
      
      'Количество дубликатов для второго региона:',len(duplicated_list_2),'\n'
      'Список дубликатов для второго региона:','\n',duplicated_list_2,'\n','\n'
      
      'Количество дубликатов для третьего региона:',len(duplicated_list_3),'\n'
      'Список дубликатов для третьего региона:','\n',duplicated_list_3,'\n')

Количество дубликатов для первого региона: 10 
Список дубликатов для первого региона: 
 7530     HZww2
41724    bxg6G
51970    A5aEY
63593    QcMuo
66136    74z30
69163    AGS9W
75715    Tdehs
90815    fiKDv
92341    TtcGQ
97785    bsk9y
Name: id, dtype: object 
 
Количество дубликатов для второго региона: 4 
Список дубликатов для второго региона: 
 41906    LHZR0
82178    bfPNe
82873    wt4Uk
84461    5ltQ6
Name: id, dtype: object 
 
Количество дубликатов для третьего региона: 4 
Список дубликатов для третьего региона: 
 41906    TzK85
82178    WzvSU
82873    uxqyU
84461    unTDF
Name: id, dtype: object 



Список `id` - дубликатов сформирован.

In [7]:
#удалим столбец 'id' из датасетов

data_1 = data_1.drop(['id'], axis = 1)
data_2 = data_2.drop(['id'], axis = 1)
data_3 = data_3.drop(['id'], axis = 1)

print(data_1.info(),'\n')
print(data_2.info(),'\n')
print(data_3.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   f0       100000 non-null  float64
 1   f1       100000 non-null  float64
 2   f2       100000 non-null  float64
 3   product  100000 non-null  float64
dtypes: float64(4)
memory usage: 3.1 MB
None 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   f0       100000 non-null  float64
 1   f1       100000 non-null  float64
 2   f2       100000 non-null  float64
 3   product  100000 non-null  float64
dtypes: float64(4)
memory usage: 3.1 MB
None 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   f0       100000 non-null  fl

Столбец `id` удален, другие данные не изменились

### **Вывод по блоку "Анализ и подготовка данных"**

В рамках данного блока были выполнены следующие процедуры:
1. Импортированы необходимые для проекта библиотеки и функции.
2. Проведена загрузка и первичный анализ данных:

Было получено 3 датасета, каждый содержал **10 000** строк и **5** столбцов. В данных отсутствуют пропуски, в колонке `id` имеются дубликаты.

3. Предобработка данных на основе анализа:
Были сформированы списоки дубликатов в в колонке `id`, так как при финальном выборе месторождения `id` лучшего местоположения может попасть в список дубликатов, что может привести к ошибке при выборе месторождения. При этом из датафреймов строки с дублирующими `id` не удалены, так как потенциально могут иметь высокие показатели.

Так как значение столбца `id` для обучения модели не является необходимым - данный элемент удален из датафреймов

## Обучение и проверка модели

На основе предобработанных данных из блока 1 проведем следующие действия:
1. Разделим популяции на `обучающую` и `валидационную` выборку для каждого региона;
2. Масштабируем параметры;
3. Обучим модель Линейной регрессии

### Разделение популяции на обучающую и валидационную выборку

В первую очередь перед разделением необходимо выделить общие признаки и целевой признак.

В данных датафреймах колонки относятся к следующим признакам:

- `f0`, `f1`, `f2` — общие признаки (три признака точек);
- `product` — целевой признак (объём запасов в скважине).

Для обучения моделей необходимо разделить датафреймы на две выборки - на одной выборке модель будет обучаться, а на основе второй будет исследовано качество предсказаний модели.

Разделим популяцию каждого региона в следующей пропорции:
- обучающая выборка - **75%**
- валидационная выборка - **25%**

Для разделения популяции воспользуемся методом `train_test_split`

In [8]:
#сформируем одинаковое значение для random_state
state = np.random.RandomState(12345)

In [9]:
#разделение для первого региона
#разделим признаки

features_1 = data_1.drop(['product'], axis = 1)
target_1 = data_1['product']

#сформируем общие размеры популяции до разделения:

print ('Размер популяции до деления: в первом регионе','\n',
       'размер общих признаков:', features_1.shape, '\n',
       'размер целевых признаков:', target_1.shape, '\n')

#проведем разделение популяции на выборки
features_train_1, features_valid_1, target_train_1, target_valid_1 = train_test_split (
    features_1, target_1, test_size = 0.25, random_state = state)

#сформируем размеры выборок после разделения:
print ('Размер популяции после деления в первом регионе:','\n',
       'размер общих обучающих признаков:', features_train_1.shape,'\n',
       'размер целевых обучающих признаков:', target_train_1.shape,'\n',
       'размер общих валидационных признаков:', features_valid_1.shape,'\n',
       'размер целевых валидационных признаков:', target_valid_1.shape)

Размер популяции до деления: в первом регионе 
 размер общих признаков: (100000, 3) 
 размер целевых признаков: (100000,) 

Размер популяции после деления в первом регионе: 
 размер общих обучающих признаков: (75000, 3) 
 размер целевых обучающих признаков: (75000,) 
 размер общих валидационных признаков: (25000, 3) 
 размер целевых валидационных признаков: (25000,)


In [10]:
#разделение для второго региона
#разделим признаки

features_2 = data_2.drop(['product'], axis = 1)
target_2 = data_2['product']

#сформируем общие размеры популяции до разделения:

print ('Размер популяции до деления: во втором регионе','\n',
       'размер общих признаков:', features_2.shape, '\n',
       'размер целевых признаков:', target_2.shape, '\n')

#проведем разделение популяции на выборки
features_train_2, features_valid_2, target_train_2, target_valid_2 = train_test_split (
    features_2, target_2, test_size = 0.25, random_state = state)

#сформируем размеры выборок после разделения:
print ('Размер популяции после деления во втором регионе:','\n',
       'размер общих обучающих признаков:', features_train_2.shape,'\n',
       'размер целевых обучающих признаков:', target_train_2.shape,'\n',
       'размер общих валидационных признаков:', features_valid_2.shape,'\n',
       'размер целевых валидационных признаков:', target_valid_2.shape)

Размер популяции до деления: во втором регионе 
 размер общих признаков: (100000, 3) 
 размер целевых признаков: (100000,) 

Размер популяции после деления во втором регионе: 
 размер общих обучающих признаков: (75000, 3) 
 размер целевых обучающих признаков: (75000,) 
 размер общих валидационных признаков: (25000, 3) 
 размер целевых валидационных признаков: (25000,)


In [11]:
#разделение для третьего региона
#разделим признаки

features_3 = data_3.drop(['product'], axis = 1)
target_3 = data_3['product']

#сформируем общие размеры популяции до разделения:

print ('Размер популяции до деления: в третьем регионе','\n',
       'размер общих признаков:', features_3.shape, '\n',
       'размер целевых признаков:', target_3.shape, '\n')

#проведем разделение популяции на выборки
features_train_3, features_valid_3, target_train_3, target_valid_3 = train_test_split (
    features_3, target_3, test_size = 0.25, random_state = state)

#сформируем размеры выборок после разделения:
print ('Размер популяции после деления в третьем регионе:','\n',
       'размер общих обучающих признаков:', features_train_3.shape,'\n',
       'размер целевых обучающих признаков:', target_train_3.shape,'\n',
       'размер общих валидационных признаков:', features_valid_3.shape,'\n',
       'размер целевых валидационных признаков:', target_valid_3.shape)

Размер популяции до деления: в третьем регионе 
 размер общих признаков: (100000, 3) 
 размер целевых признаков: (100000,) 

Размер популяции после деления в третьем регионе: 
 размер общих обучающих признаков: (75000, 3) 
 размер целевых обучающих признаков: (75000,) 
 размер общих валидационных признаков: (25000, 3) 
 размер целевых валидационных признаков: (25000,)


**Вывод по пункту 2.1**

Популяции для каждого региона разделены на обучающую и валидационную в пропорции **75/25** соответственно.

Разделение прошло корректно, все данные разделены в верной пропорции.

### Масштабирование параметров

Модели линейной регрессии очень сильно реагируют на данные, в рамках одного признака, слишком отличающиеся от других - в таком случае модель присваивает более высоким признакам более высокий вес, а это не является корректным. Чтобы избежать данной ловушки нам необходимо "отмасштабировать" данные - привести к одному масштабу.

Произвести масштабирование данных можно с помощью SatndardScaler()

In [12]:
#убираем ошибку от применения StandardScaler
pd.options.mode.chained_assignment = None

#укажем числовые переменные
numeric = ['f0', 'f1', 'f2']

In [13]:
#для первого региона
#отмасштабируем данные 
scaler_1 = StandardScaler()
scaler_1.fit(features_train_1[numeric])

features_train_1[numeric] = scaler_1.transform (features_train_1[numeric])
features_valid_1[numeric] = scaler_1.transform (features_valid_1[numeric])

In [14]:
#для первого региона
#отмасштабируем данные 
scaler_2 = StandardScaler()
scaler_2.fit(features_train_2[numeric])

features_train_2[numeric] = scaler_2.transform (features_train_2[numeric])
features_valid_2[numeric] = scaler_2.transform (features_valid_2[numeric])

In [15]:
#для первого региона
#отмасштабируем данные 
scaler_3 = StandardScaler()
scaler_3.fit(features_train_3[numeric])

features_train_3[numeric] = scaler_3.transform (features_train_3[numeric])
features_valid_3[numeric] = scaler_3.transform (features_valid_3[numeric])

**Вывод по пункту 2.2**

Данные отмасштабированы.

### Обучим модель Линейной регрессии 

Согласно условиям задания, для обучения лучше использовать модель Линейной регрессии. Из-за малого количества гиперпараметров линейная регрессия менее склонна к переобучению, чем деревья решений.

Наиболее распространённая метрика качества в задаче регрессии — **средняя квадратичная ошибка(MSE)**. однако так как MSE - число возведенное в квадрат, определим метрику **корень из средней квадратичной ошибки (RMSE)**.

Метрика **RMSE** показывает как сильно отклоняется предсказанное значение от реального - таким образом, чем **RMSE** меньше, тем лучше предсказывает модель.

Сделаем предсказания на основе валидационной выборки и сравним с фактическими целевыми значениями валидационной выборки.

In [16]:
#для первого региона
model_1 = LinearRegression()
model_1.fit (features_train_1, target_train_1)
predicted_valid_1 = model_1.predict(features_valid_1)
rmse_1 = mean_squared_error(target_valid_1, predicted_valid_1)** 0.5


print('Средний запас предсказанного сырья для первого региона:', predicted_valid_1.mean(), '\n' 
      'Средний запас фактического сырья для первого региона:', target_valid_1.mean(), '\n'
      'RMSE для первого региона', rmse_1)

Средний запас предсказанного сырья для первого региона: 92.59256778438035 
Средний запас фактического сырья для первого региона: 92.07859674082927 
RMSE для первого региона 37.5794217150813


In [17]:
#для второго региона
model_2 = LinearRegression()
model_2.fit (features_train_2, target_train_2)
predicted_valid_2 = model_2.predict(features_valid_2)
rmse_2 = mean_squared_error(target_valid_2, predicted_valid_2)** 0.5


print('Средний запас предсказанного сырья для второго региона:', predicted_valid_2.mean(), '\n' 
      'Средний запас фактического сырья для второго региона:', target_valid_2.mean(), '\n'
      'RMSE для третьего региона', rmse_2)

Средний запас предсказанного сырья для второго региона: 68.76995145799754 
Средний запас фактического сырья для второго региона: 68.77162424984647 
RMSE для третьего региона 0.889736773768065


In [18]:
#для третьего региона
model_3 = LinearRegression()
model_3.fit (features_train_3, target_train_3)
predicted_valid_3 = model_3.predict(features_valid_3)
rmse_3 = mean_squared_error(target_valid_3, predicted_valid_3)** 0.5


print('Средний запас предсказанного сырья для третьего региона:', predicted_valid_3.mean(), '\n' 
      'Средний запас фактического сырья для третьего региона:', target_valid_3.mean(), '\n'
      'RMSE для третьего региона', rmse_3)

Средний запас предсказанного сырья для третьего региона: 95.087528122523 
Средний запас фактического сырья для третьего региона: 94.74895871720241 
RMSE для третьего региона 39.958042459521614


**Вывод по пункту 2.3**

Модели Линейной регрессии обучены, предсказания моделей на валидационных данных сформулированы. 

### **Вывод по блоку "Обучение и проверка модели"**

В рамках данного блока были получены следующие результаты:

1. Популяции, для каждого региона были разделены на две выборки в следующей пропорции:
- обучающая выборка - **75%**
- валидационная выборка - **25%**

2. С помощью метода `Масштабирвоания` все признаки были приведены к общему масштабу с целью избежания слишком высоких или слишком низких значений признаков, которые могли повлиять на качество модели.

3. Для каждого региона мы обучили модель Линейной регрессии, которая по синтетическим данным определяет количество запасов сырья у месторождения. Также для каждого региона была определена метрика **RMSE** - корень из квадратного среднего отклонения.

Были получены следующие результаты:

- Для региона №1, среднее количество запасов предсказанных моделью составляет **92,59** тыс. баррелей, что достаточно близко к фактическому значению среднего количества запасов - **92.08** тыс. баррелей. При этом показатель метрики RMSE составляет **37.58**, таким образом отклонение предсказаний от факта составляет **37.58** тыс. баррелей. Таким образом при близких средних и высоком RMSE можно сделать вывод, что индивидуально предсказания и факт расходятся как в большую так и в меньшую сторону.
- Для региона №2, среднее количество запасов предсказанных моделью составляет **68,77** тыс. баррелей, что эквивалентно фактическому значению среднего количества запасов - **68.77** тыс. баррелей. При этом показатель метрики RMSE составляет **0.88**, модель делает достаточно точные предсказания.
- Для региона №3 , среднее количество запасов предсказанных моделью составляет **95,08** тыс. баррелей, что достаточно близко к фактическому значению среднего количества запасов - **94.75** тыс. баррелей. При этом показатель метрики RMSE составляет **39.96**, таким образом ситуация подобна региону один, только в данном случае отклонение еще выше.

Подводя итог можно сказать, что в Регионе 1 и Регионе 2 - достаточно высокие запасы сырья, но при этом вероятность отклонения достаточно велика, в Регионе 2 - запасов меньше (примерно на 25%) - однако модель достаточно точно предсказывает значения каждого месторождения.

## Подготовка к расчёту прибыли

В данном блоке будут подготовлены данные для расчета прибыли по каждому месторождению:

1. Сформированы константы, которые будут участвовать в расчетах;
2. Рассчитана точка безубыточности для бурения новой скважины. 
3. Объем сырья для точки безубыточности будет сравнен со средними значениями сырья для каждого региона.
 

### Константы для расчетов

Имеются вводные данные:
- При разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки.
- Бюджет на разработку скважин в регионе — 10 млрд рублей.
- При нынешних ценах один баррель сырья приносит 450 рублей дохода. Доход с каждой единицы продукта составляет 450 тыс. рублей, поскольку объём указан в тысячах баррелей.
- После оценки рисков нужно оставить лишь те регионы, в которых вероятность убытков меньше 2.5%.

Так как количество сырья указано в тысячах рублей, для сохранения масштаба все монетарные переменные будут в тысячах рублей

Сохраним эту информацию в переменных

In [19]:
#константу бюджета сохраним в переменную 'BUDGET' - тыс. руб.:
BUDGET = 10000000
#прибыль на одну тысячу баррелей сохраним в переменную 'PFB' - profit_from_barrel - тыс. руб:
PFB = 450
#количество исследуемых месторождений 'NED' - number explored deposits - шт.
NED = 500
#количество месторождения для разработки 'QFD' - quantity field development - шт.
QFD = 200


**Вывод по пункту 3.2:**

Константы сформированы.

### Точка безубыточности

Точка безубыточности - объём производства и реализации продукции, при котором расходы будут компенсированы доходами. В рамках данного проекта под *объемом* подразумевается - запас сырья. Таким образом точкой безубыточности будет являться такое количество сырья, расходы на добычу которого будут компенсированы доходами от его реализации.

Объем добытый сверх точки безубыточности является чистой прибылью организации.

Рассчитаем точку безубыточности для данного проекта и сохраним в переменную `BEP` - break-even point.

In [20]:
#расчет происходит по формуле бюджет/стоимость единицы
BEP = BUDGET/PFB

print ('Объем сырья, который необходимо добыть в регионе составляет:',BEP)

Объем сырья, который необходимо добыть в регионе составляет: 22222.222222222223


**Вывод по пункту 3.2:**

Для того, чтобы покрыть все расходы на разработку необходимо добыть в регионе не менее **22 223** тыс. баррелей сырья.

### Сравнение точки безубыточности с запасами в регионе

Так как по условиям проекта будет разработано 200 месторождений в выбранном регионе, необходимо для каждого региона умножить среднее фактическое значение (рассчитанное выше) на константу `qfd` (которая хранит в себе те самые 200 шт.). Полученное значение необходимо сравнить с точкой безубыточности.

In [21]:
# сравнение средней добычи и точки безубыточности

print('Точка безубыточности по региону:', BEP,'\n','\n'
      'Размер средней добычи в первом регионе:', target_valid_1.mean()*QFD,'\n'
      'Размер средней добычи во втором регионе:',target_valid_2.mean()*QFD,'\n'
      'Размер средней добычи в третьем регионе:',target_valid_3.mean()*QFD)

Точка безубыточности по региону: 22222.222222222223 
 
Размер средней добычи в первом регионе: 18415.719348165854 
Размер средней добычи во втором регионе: 13754.324849969293 
Размер средней добычи в третьем регионе: 18949.79174344048


**Вывод по блоку 3.3**

При анализе было выявлено, что если мы будем брать средние значения сырья в регионе - мы не выйдем на доход, который в состоянии компенсировать расходы на добычу. Однако в данном проекте для разведки будут отобраны месторождения с наиболее высокими показателями запасов сырья.

### **Вывод по блоку "Подготовка к расчету прибыли"**

В рамках данного блока были выполнены следующие задачи:
1. Сформированы константы, необходимые для расчетов, были выделены следующие константы:
- **BUDGET** = 10000000 - бюджет разработки;
- **PFB** = 450 - прибыль с 1 тыс. баррелей сырья
- **NED** = 500 - количество исследуемых месторождений
- **QFD** = 200 - лучшие месторождения из исследуемых
- **TH** = 0.025 - критический уровень убытков

2. Была определена точка безубыточности - для того чтобы разработка была прибыльной, в регионе необходимо добыть не менее **22223** тыс. баррелей сырья.

3. Было проведено сравнение со средними запасами сырья в каждом регионе, для этого среднее значения сырья одного месторождения умножили на количество разрабатываемых месторождений (согласно плану) - 200 шт.
Значения для регионов:
- Размер средней добычи в первом регионе: **18415.71** 
- Размер средней добычи во втором регионе: **13754.32** 
- Размер средней добычи в третьем регионе: **18949.79**

Для каждого региона при средней добычи и требуемой точкой безубыточности в 22222 тыс. баррелей сырья - проект разработки будет убыточен, однако планируется отбирать лучшие варианты. Детально см. следующий блок.

## Расчёт прибыли и рисков 

В данном блоке будут выполнены следующие операции:

1. Построена функция определяющая прибыль по каждому региону;
2. Проведена процедура *Bootstrap* 
3. Проведен выбор лучшего региона.

### Функция расчета прибыли

Согласно условиям для определения прибыли будут исследоваться 500 месторождений и из них будут отбираться 200 месторождений с лучшими показателями согласно предсказанию модели. 

Таким образом в функцию будут поступать данные о 500 месторождениях - далее будет фильтроваться результаты предсказания в порядке убывания и отсекаться 200 (`QFD`) строк.

Для определения прибыли необходимо будет умножить сумму запасов двухста месторождений на стоимость 1 тыс. баррелей (`PFB`) и вычесть планируемый бюджет разработки месторождений (`BUDGET`)

Построенная функция представлена ниже:

In [22]:
def revenue (target, predictions):
    predictions_sorted = predictions.sort_values(ascending = False)
    selected = target[predictions_sorted.index][:QFD]   
    return selected.sum()*PFB-BUDGET

### Применение процедуры Bootstrap

Bootstrap - процедура которая семплирует данные согласно указанным гипперпараметрам.

Согласно услвоиям функция будет переберать выборки **1000** раз, каждый раз семплируюя случайные позиции. 

Гиперпараметрами данной функции будут являться значения:
- n - число элементов в выборке - 500 шт.`NED`
- random_state = 12345 
- replace = True - данный гиперпараметр позволяет возвращать в выборку позиции, которые уже были использованы в других выборках.

Результатом функции будет список из **500** целевых признаков и предсказаний.

Дополнительно добавим функцию `bts`, которая будет принимать значения целевых признаков и предсказаний.

In [23]:
 def bts (target, predictions):
    
    values = []
    loss =0
    for i in tqdm(range(1000)):
        target_subsample = (target.reset_index(drop=True)
                            .sample(n = NED, random_state = state, replace = True))
        predictions_subsample = pd.Series(predictions)[target_subsample.index]
        
#Данный блок необходим для расчета риска убытков
        profit = revenue(target_subsample, predictions_subsample)
        if profit <0:
            loss += 1
        
        values.append (revenue(target_subsample, predictions_subsample))
        
    values = pd.Series(values)
    
#расчет %% убытков и вывод значений    
    loss_prob = loss/1000*100
    return values, loss_prob

### Выбор лучшего региона

Для определения лучшего региона нам необходимо получить информацию о следующих показателях добычи в данном регионе:
- средняя прибыль региона;
- 95%-й доверительный интервал;
- риск убытков.

Рассчитаем необходимые значения для каждого региона

In [24]:
region_1, prob_loss_region_1 = bts(target_valid_1, predicted_valid_1)
print ('Информация для региона №1', '\n','\n'
       'Средняя прибыль в регионе составляет, тыс. руб.: ',region_1.mean(), '\n','\n'
       'Значения 95% доверительного интервала между: ', '\n',
       'Нижнее значение: ', region_1.quantile(0.025), '\n',
       'Верхнее значение: ', region_1.quantile(1-0.025),'\n','\n'
       'Риск убытков, %: ', prob_loss_region_1) 

100%|██████████| 1000/1000 [00:02<00:00, 389.24it/s]

Информация для региона №1 
 
Средняя прибыль в регионе составляет, тыс. руб.:  423897.2379169053 
 
Значения 95% доверительного интервала между:  
 Нижнее значение:  -76187.81389036294 
 Верхнее значение:  957846.5319517821 
 
Риск убытков, %:  4.8


In [25]:
region_2, prob_loss_region_2 = bts(target_valid_2, predicted_valid_2)
print ('Информация для региона №2', '\n','\n'
       'Средняя прибыль в регионе составляет, тыс. руб.: ',region_2.mean(), '\n','\n'
       'Значения 95% доверительного интервала между: ', '\n',
       'Нижнее значение: ', region_2.quantile(0.025), '\n',
       'Верхнее значение: ', region_2.quantile(1-0.025),'\n','\n'
       'Риск убытков, %: ', prob_loss_region_2) 

100%|██████████| 1000/1000 [00:02<00:00, 363.44it/s]

Информация для региона №2 
 
Средняя прибыль в регионе составляет, тыс. руб.:  513256.69891726074 
 
Значения 95% доверительного интервала между:  
 Нижнее значение:  108066.89523396171 
 Верхнее значение:  928574.439232496 
 
Риск убытков, %:  0.6


In [26]:
region_3, prob_loss_region_3 = bts(target_valid_3, predicted_valid_3)
print ('Информация для региона №3', '\n','\n'
       'Средняя прибыль в регионе составляет, тыс. руб.: ',region_3.mean(), '\n','\n'
       'Значения 95% доверительного интервала между: ', '\n',
       'Нижнее значение: ', region_3.quantile(0.025), '\n',
       'Верхнее значение: ', region_3.quantile(1-0.025),'\n','\n'
       'Риск убытков, %: ', prob_loss_region_3) 

100%|██████████| 1000/1000 [00:02<00:00, 372.00it/s]

Информация для региона №3 
 
Средняя прибыль в регионе составляет, тыс. руб.:  381120.3595759014 
 
Значения 95% доверительного интервала между:  
 Нижнее значение:  -142800.63008786933 
 Верхнее значение:  893380.5657504003 
 
Риск убытков, %:  7.3999999999999995


### **Вывод по блоку "Расчет прибыли и рисков"**

В данном блоке были выполнены следующие процедуры:

1. Сформирована функция для расчета прибыли в регионе;
2. Сформирована функция для применения процедуры bootstrap
3. Для каждого региона были получены показатели:

**Регион №1**

Средняя прибыль в регионе составляет:  **425 063.85** тыс. руб.
 
Значения 95% доверительного интервала между:  
 Нижнее значение:  -120224.31621769907 
 Верхнее значение:  982554.8633446723 
 
Таким образом с вероятностью 95% прибыль в регионе будет находится в интервале от **-120 224.31** до **982 554.86**
 
Риск убытков:  **6.2%** - что противоречит заданному уровню риска в 2,5%

**Регион №2**

Средняя прибыль в регионе составляет:  **516 331.42** тыс. руб.
 
Значения 95% доверительного интервала между:  
 Нижнее значение:  117121.63594472969 
 Верхнее значение:  960002.4599272121 
 
Таким образом с вероятностью 95% прибыль в регионе будет находится в интервале от **117 121.63** до **960 002.45**
 
Риск убытков:  **0.5%** - что допускается заданным уровнем риска в 2,5%

**Регион №3**

Средняя прибыль в регионе составляет:  **364 601.99** тыс. руб.
 
Значения 95% доверительного интервала между:  
 Нижнее значение:  -161963.86069313608 
 Верхнее значение:  929509.7313283813 
 
Таким образом с вероятностью 95% прибыль в регионе будет находится в интервале от **-161963.86** до **929509.73**
 
Риск убытков:  **9.7%** - что противоречит заданному уровню риска в 2,5%

## Вывод по проекту

Целью данного проекта является сформировать понимание, в каком из трех предложенных регионов стоит начать разработку месторождений, необходимо выбрать самый прибыль регион с наименьшем уровнем риска убытков.

**Блок №1 - Анализ и подготовка данных**

В рамках первого блока была проведена загрузка и предварительный анализ данных по трем датафреймам содержащим данные в отношении запасов сырья и дополнительных параметрах расположенных в регионе месторождений в разрезе индивидуального месторождения. В каждом датафрейме имеется **10 000** строк и **5** признаков - где одна строка - одно месторождение.

В ходе процедуры анализа были выявлены дубликаты, информация о которых сохранена в отдельный список. Данная информация необходима для того чтобы не допустить ошибку при выборе месторождения для разработки.

Также в ходе предобработки был удален столбец содержащий информация о `ID` месторождения, так как данная информация не нужна для обучения модели.

**Блок №2 - Обучение и проверка модели**

В данном блоке были выполнены шаги для успешного обучения моделей прогнозирования.

Популяции по каждому региону были разделены на 2 выборки - обучающую и валидационную в пропорции **75/25** соответственно.

Далее признаки на которых обучается модель были приведены к одному масштабу с помощью функции `StandardScaler`, с целью уменьшения ошибки при некорректном присвоении веса.

После этого были проведено обучение моделей `Линейно регрессии` для каждого региона. В качестве метрики качества предсказания моделей была использована метрика **RMSE** - которая показывает насколько сильно среднее значение предсказания отличается от фактического значения. Качество предсказаний моделей составляют следующее значения:
- качество модели первого региона составляет: 37.58 тыс. баррелей
- качество модели второго региона составляет: 0.88 тыс. баррелей
- качество модели третьего региона составляет: 39.96 тыс. баррелей

**Блок №3 - Подготовка к расчету прибыли**

Блок №3 затрагивает формирование информации для расчета прибыли, в этом блоке были выполнены операции:
1. Формирование констант для использования в расчетах;
2. Определена точка безубыточности для добычи в регионе, которая составили **22 223** тыс. баррелей.
3. Определено среднее количество добываемого сырья в каждом из регионов. Для этого среднее значение сырья для одного месторождения умножили на 200 (Число планируемых скважин в регионе). Для регионов были получены следующие средние показатели добычи:
- Размер средней добычи в первом регионе: **18415.71** 
- Размер средней добычи во втором регионе: **13754.32** 
- Размер средней добычи в третьем регионе: **18949.79**

Согласно полученным данных есть риск, что добыча не будет успешной, так как ни один из регионов не показал необходимой цифры. Однако согласно условиям с помощью моделей машинного обучения мы отберем наиболее богатые на сырье месторождения.

**Блок №4 - Расчет прибыли и рисков**

В данном блоке были проведены основные расчеты:
- Сформирована функция отвечающая за расчет прибыли;
- Сформирована функция в рамках которой применяется процедура Bootstrap;
- Рассчитаны основные показатели для каждого региона.

Для каждого региона были получены следующие значения:

Потенциальная прибыль от разработки в регионе:
- Регоин №1: **425 063.85** тыс. руб
- Регоин №2: **516 331.42** тыс. руб.
- Регоин №3: **364 601.99** тыс. руб.

95% доверительный интервал для каждого региона:
- Регоин №1: от **-120 224.31** до **982 554.86**
- Регоин №2: от **117 121.63** до **960 002.45**
- Регоин №3: от **-161963.86** до **929509.73**

Потенциальные убытки от разработки в каждом регионе:
- Регоин №1: **6.2%**
- Регоин №2: **0.5%**
- Регоин №3: **9.7%**


**Рекомендации**

На основе полученных данных можно сделать вывод, что самым прибыльным регионом для запуска проекта по разработке месторождений является **Регион №2** - средняя прогнозная прибыль составляет **516 331.42**. При этом, согласно входящим данным риск убытка от разработки не должен составлять более **2.5%** - значения риска для второго региона - **0.5**%. Таким образом **Регион №2** лучше всего подходит для запуска проекта по разработке и добычи сырья.

Что касается Региона №1 и Региона №3 - несмотря на прибыльность в данных регионах, уровень риска убытка составляет **6.2%** и **9.7%** - что противоречит установленному уровню допустимого риска.